cli/azd/pkg/azapi/stack_deployments.go

```diff
@@ -0,0 +1,628 @@
+		// If a deployment stack is not found with the given name, fallback to check for standard deployments
+		if errors.Is(err, ErrDeploymentNotFound) {
+			return d.standardDeployments.GetSubscriptionDeployment(ctx, subscriptionId, deploymentName)
+		}
+
+		return nil, err
+	}
+
+	return d.convertFromStackDeployment(deploymentStack), nil
+}
+
+func (d *StackDeployments) ListResourceGroupDeployments(
+	ctx context.Context,
+	subscriptionId string,
+	resourceGroupName string,
+) ([]*ResourceDeployment, error) {
+	client, err := d.createClient(ctx, subscriptionId)
+	if err != nil {
+		return nil, err
+	}
+
+	results := []*ResourceDeployment{}
+
+	pager := client.NewListAtResourceGroupPager(resourceGroupName, nil)
+	for pager.More() {
+		page, err := pager.NextPage(ctx)
+		if err != nil {
+			return nil, err
+		}
+
+		for _, deployment := range page.Value {
+			results = append(results, d.convertFromStackDeployment(deployment))
+		}
+	}
+
+	return results, nil
+}
+
+func (d *StackDeployments) GetResourceGroupDeployment(
+	ctx context.Context,
+	subscriptionId string,
+	resourceGroupName string,
+	deploymentName string,
+) (*ResourceDeployment, error) {
+	client, err := d.createClient(ctx, subscriptionId)
+	if err != nil {
+		return nil, err
+	}
+
+	var deploymentStack *armdeploymentstacks.DeploymentStack
+
+	err = retry.Do(
+		ctx,
+		retry.WithMaxDuration(10*time.Minute, retry.NewConstant(5*time.Second)),
+		func(ctx context.Context) error {
+			response, err := client.GetAtResourceGroup(ctx, resourceGroupName, deploymentName, nil)
+			if err != nil {
+				return fmt.Errorf(
+					"%w: '%s' in resource group '%s', Error: %w",
+					ErrDeploymentNotFound,
+					resourceGroupName,
+					deploymentName,
+					err,
+				)
+			}
+
+			if response.DeploymentStack.Properties.DeploymentID == nil {
+				return retry.RetryableError(errors.New("deployment stack is missing ARM deployment id"))
+			}
+
+			deploymentStack = &response.DeploymentStack
+
+			return nil
+		})
+
+	if err != nil {
+		// If a deployment stack is not found with the given name, fallback to check for standard deployments
+		if errors.Is(err, ErrDeploymentNotFound) {
+			return d.standardDeployments.GetResourceGroupDeployment(ctx, subscriptionId, resourceGroupName, deploymentName)
+		}
+
+		return nil, err
+	}
+
+	return d.convertFromStackDeployment(deploymentStack), nil
+}
+
+func (d *StackDeployments) DeployToSubscription(
+	ctx context.Context,
+	subscriptionId string,
+	location string,
+	deploymentName string,
+	armTemplate azure.RawArmTemplate,
+	parameters azure.ArmParameters,
+	tags map[string]*string,
+) (*ResourceDeployment, error) {
+	client, err := d.createClient(ctx, subscriptionId)
+	if err != nil {
+		return nil, err
+	}
+
+	templateHash, err := d.CalculateTemplateHash(ctx, subscriptionId, armTemplate)
+	if err != nil {
+		return nil, fmt.Errorf("failed to calculate template hash: %w", err)
+	}
+
+	tags[azure.TagKeyAzdDeploymentTemplateHashName] = &templateHash
```

Comment:
In standard deployments the template hash is available directly on the deployment objects and available during list operations like list sub/rg deployments.  

However - for stack deployments we do know the deployment id associated with the stack but would need to do a separate GET operation for each deployment stack to get the template hash.

Instead of waiting for an additional N+1 operations I am caching it on the deployment stack in the tags since this metadata is not directly available on the list deployment stack operations.

Reply:
Sounds good - I'd like to avoid the N+1 behavior as well.  Maybe over time we can figure out a way to structure things such that that isn't a problem (like defer the work until it is just needed) but this is a good optimization for the world we live in today.